README.md
[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Total Downloads][ico-downloads]][link-downloads]

# powerdns-php
A PHP client to communicate with the PowerDNS API.

## Install
Via Composer

```bash
$ composer require exonet/powerdns-php
```

## Usage
Basic example how to create a new DNS zone and insert a few DNS records.
```php
use Exonet\Powerdns\Powerdns;
use Exonet\Powerdns\RecordType;

// Initialize the Powerdns client.
$powerdns = new Powerdns('127.0.0.1', 'powerdns_secret_string');

// Create a new zone.
$zone = $powerdns->createZone(
    'example.com',
    ['ns1.example.com.', 'ns2.example.com.']
);

// Add two DNS records to the zone.
$zone->create([
    ['type' => RecordType::A, 'content' => '127.0.0.1', 'ttl' => 60, 'name' => '@'],
    ['type' => RecordType::A, 'content' => '127.0.0.1', 'ttl' => 60, 'name' => 'www'],
]);
```

See the [examples](examples) directory for more.

## Change log
Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing
Testing against multiple PHP / PowerDNS versions can be done by using the provided `docker-compose.yml` and the
`run-tests.sh` shell script:

``` bash
$ docker-compose up -d
$ ./run-tests.sh
```

After running `docker-compose up -d` wait a few seconds so PowerDNS can be initialized. You can leave the containers
running and call the test script multiple times.

To test against a specific PHP / PowerDNS combination, you can provide the PHP version as first and the PowerDNS
version as second parameter:

```bash
$ ./run-tests.sh 7.4 4.3
```

## Contributing
Please see [CONTRIBUTING](.github/CONTRIBUTING.md) and [CODE_OF_CONDUCT](.github/CODE_OF_CONDUCT.md) for details.

## Security
If you discover any security related issues, please email development@exonet.nl instead of using the issue tracker.

## Credits
- [Exonet][link-author]
- [All Contributors][link-contributors]

## License
The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/exonet/powerdns-php.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/com/exonet/powerdns-php/master.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/exonet/powerdns-php.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/exonet/powerdns-php
[link-travis]: https://travis-ci.com/exonet/powerdns-php
[link-downloads]: https://packagist.org/packages/exonet/powerdns-php
[link-author]: https://github.com/exonet
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `powerdns-php` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## Unreleased
[Compare v3.4.0 - Unreleased](https://github.com/exonet/powerdns-php/compare/v3.4.0...develop)

## [v3.4.0](https://github.com/exonet/powerdns-php/releases/tag/v3.4.0) - 2021-10-19
[Compare v3.3.1 - v3.4.0](https://github.com/exonet/powerdns-php/compare/v3.3.1...v3.4.0)
### Added
- The `setConnector` method in the Powerdns class.

### Changed
- The private methods in the `Connector` class are now protected instead of private.

### Fixed
- Filling the `SearchResult` resource no longer results in 'Undefined index' notices. (#78)

## [v3.3.1](https://github.com/exonet/powerdns-php/releases/tag/v3.3.1) - 2021-07-06
[Compare v3.3.0 - v3.3.1](https://github.com/exonet/powerdns-php/compare/v3.3.0...v3.3.1)
### Fixed
- Replace deprecated Guzzle function (#75)

## [v3.3.0](https://github.com/exonet/powerdns-php/releases/tag/v3.3.0) - 2021-06-07
[Compare v3.2.0 - v3.3.0](https://github.com/exonet/powerdns-php/compare/v3.2.0...v3.3.0)
### Added
- Added interfaces for Connector and Powerdns. (@frankvanhest - #72)
- Run tests against PowerDNS 4.4. (@frankvanhest - #73)

## [v3.2.0](https://github.com/exonet/powerdns-php/releases/tag/v3.2.0) - 2021-05-06
[Compare v3.1.0 - v3.2.0](https://github.com/exonet/powerdns-php/compare/v3.1.0...v3.2.0)
### Added
- Added a method `unsetNsec3param()` to unset the nsec3param for a zone.
- Unit tests for enabling, disabling and setting DNSSEC.
- It is now possible to set/get resource records to a zone resource. See `examples/new_domain_from_zone_resource.php` for an example.

### Changed
- Calling `setNsec3Param()` with value `null` will unset the `nsec3param`.
- The internals of getting ResourceSets. Previously a GET request was made to get the resource sets, now the zone resource is being used which already contains all resource sets.

### Fixed
- Updated the stable composer.phar URL to properly run tests.

## [v3.1.0](https://github.com/exonet/powerdns-php/releases/tag/v3.1.0) - 2021-02-01
[Compare v3.0.0 - v3.1.0](https://github.com/exonet/powerdns-php/compare/v2.0.0...v3.1.0)
### Added
- The `omitDnssecAndEditedSerialFields` parameter in `listZones`. (@frankvanhest - #61)

## [v3.0.0](https://github.com/exonet/powerdns-php/releases/tag/v3.0.0) - 2021-01-05
[Compare v2.6.0 - v3.0.0](https://github.com/exonet/powerdns-php/compare/v2.6.0...v3.0.0)
### Breaking
- Dropped support for unsupported PHP versions. PHP 7.3 (or newer) is now required.

### Changed
- Updated Guzzle to 7.2. (@itspluxstahre #50)
- Updated PHPUnit to 9.5.0.
- When running tests a `composer update` will be executed for each environment to validate the packages/config for the specified PHP version.

## [v2.6.0](https://github.com/exonet/powerdns-php/releases/tag/v2.6.0) - 2020-12-21
[Compare v2.5.2 - v2.6.0](https://github.com/exonet/powerdns-php/compare/v2.5.2...v2.6.0)
### Added
- Support for PHP 8. (@akondas #54)
- Support for the PowerDNS search endpoint. (#51)
- The `notify` method on a zone. (@frankvanhest #56)

## [v2.5.2](https://github.com/exonet/powerdns-php/releases/tag/v2.5.2) - 2020-11-26
[Compare v2.5.1 - v2.5.2](https://github.com/exonet/powerdns-php/compare/v2.5.1...v2.5.2)
### Added
- The `LUA` record type. (#53)

## [v2.5.1](https://github.com/exonet/powerdns-php/releases/tag/v2.5.1) - 2020-11-06
[Compare v2.5.0 - v2.5.1](https://github.com/exonet/powerdns-php/compare/v2.5.0...v2.5.1)
### Fixed
- The `account` is send to PowerDNS when creating a zone. (#47)

## [v2.5.0](https://github.com/exonet/powerdns-php/releases/tag/v2.5.0) - 2020-10-30
[Compare v2.4.0 - v2.5.0](https://github.com/exonet/powerdns-php/compare/v2.4.0...v2.5.0)
### Added
- Get PowerDNS statistics.

## [v2.4.0](https://github.com/exonet/powerdns-php/releases/tag/v2.4.0) - 2020-08-04
[Compare v2.3.1 - v2.4.0](https://github.com/exonet/powerdns-php/compare/v2.3.1...v2.4.0)
### Added
- Export a zone in AXFR format. (#35)
- Get the zone from a ResourceRecord. (#40)
- Get the PowerDNS version of a server with `$client->serverVersion()`.
- Added a `docker-compose.yml` and `run-tests.sh` script to easy test against a combination of different PHP and PowerDNS versions.

### Changed
- Travis now uses Docker to run the tests.

### Fixed
- `$resouceRecord->getComments()` no longer throws an exception if there are no comments. (#38)

## [v2.3.1](https://github.com/exonet/powerdns-php/releases/tag/v2.3.1) - 2020-05-12
[Compare v2.3.0 - v2.3.1](https://github.com/exonet/powerdns-php/compare/v2.3.0...v2.3.1)
### Fixed
- No longer throw an exception when throwing an exception if the response body is not an error. (#32)

## [v2.3.0](https://github.com/exonet/powerdns-php/releases/tag/v2.3.0) - 2020-04-30
[Compare v2.2.0 - v2.3.0](https://github.com/exonet/powerdns-php/compare/v2.2.0...v2.3.0)
### Added
- The record type `SPF`. ([Th3Mouk](https://github.com/Th3Mouk) - [#30](https://github.com/exonet/powerdns-php/pull/30))

### Fixed
- Removed record types `APL`, `HIP` and `TA` which are not supported by PowerDNS.

## [v2.2.0](https://github.com/exonet/powerdns-php/releases/tag/v2.2.0) - 2020-04-01
[Compare v2.1.0 - v2.2.0](https://github.com/exonet/powerdns-php/compare/v2.1.0...v2.2.0)
### Added
- Create a new zone based on a zone resource object.

## [v2.1.0](https://github.com/exonet/powerdns-php/releases/tag/v2.1.0) - 2020-03-11
[Compare v2.0.0 - v2.1.0](https://github.com/exonet/powerdns-php/compare/v2.0.0...v2.1.0)
### Added
- It is now possible to set the NSEC3PARAM on a zone. ([mkevenaar-docktera](https://github.com/mkevenaar-docktera) - #25 #26)

### Changed
- The validation of the NSEC3PARAM is improved. (#27)

## [v2.0.0](https://github.com/exonet/powerdns-php/releases/tag/v2.0.0) - 2020-01-29
[Compare v1.1.0 - v2.0.0](https://github.com/exonet/powerdns-php/compare/v1.1.0...v2.0.0)
### Breaking

src/CliLogger.php
<?php

namespace Exonet\PowerDns;

use Psr\Log\AbstractLogger;

/**
 * Based on:
 * https://gist.github.com/sallar/5257396
 */
class CliLogger extends AbstractLogger
{
    public static $foreground_colors = [
        'bold' => '1',
        'dim' => '2',
        'black' => '0;30',
        'dark_gray' => '1;30',
        'blue' => '0;34',
        'light_blue' => '1;34',
        'green' => '0;32',
        'light_green' => '1;32',
        'cyan' => '0;36',
        'light_cyan' => '1;36',
        'red' => '0;31',
        'light_red' => '1;31',
        'purple' => '0;35',
        'light_purple' => '1;35',
        'brown' => '0;33',
        'yellow' => '1;33',
        'light_gray' => '0;37',
        'white' => '1;37',
        'normal' => '0;39',
    ];

    public static $background_colors = [
        'black' => '40',
        'red' => '41',
        'green' => '42',
        'yellow' => '43',
        'blue' => '44',
        'magenta' => '45',
        'cyan' => '46',
        'light_gray' => '47',
    ];

    public static $options = [
        'underline' => '4',
        'blink' => '5',
        'reverse' => '7',
        'hidden' => '8',
    ];

    public static $color = [
        'emergency' => ['foreground' => 'white', 'background' => ['red', 'underline']],
        'alert' => ['foreground' => 'white', 'background' => ['red']],
        'critical' => ['foreground' => 'red'],
        'error' => ['foreground' => 'light_red'],
        'warning' => ['foreground' => 'yellow'],
        'notice' => ['foreground' => 'normal'],
        'info' => ['foreground' => 'dark_gray'],
        'debug' => ['foreground' => 'dim'],
    ];

    /**
     * Catches static calls (Wildcard)
     *
     * @param  string $foreground_color Text Color
     * @param  array  $args             Options
     *
     * @return string                   Colored string
     */
    public static function __callStatic($foreground_color, $args)
    {
        [$string, $background] = $args;

        $colored_string = '';

        // Check if given foreground color found
        if (isset(self::$foreground_colors[$foreground_color])) {
            $colored_string .= "\033[".self::$foreground_colors[$foreground_color].'m';
        } else {
            die($foreground_color.' not a valid color');
        }

        if (is_array($background)) {
            foreach ($background as $option) {
                // Check if given background color found
                if (isset(self::$background_colors[$option])) {
                    $colored_string .= "\033[".self::$background_colors[$option].'m';
                } elseif (isset(self::$options[$option])) {
                    $colored_string .= "\033[".self::$options[$option].'m';
                }
            }
        }

        // Add string and end coloring
        $colored_string .= $string."\033[0m";

        return $colored_string;
    }

    /**
     * Logs with an arbitrary level.
     *
     * @param mixed  $level
     * @param string $message
     * @param array  $context
     */
    public function log($level, $message, array $context = []) : void
    {
        $foreground = self::$color[$level]['foreground'];
        $background = self::$color[$level]['background'] ?? null;
        $level = strtoupper($level);
        $padding = str_repeat(' ', 10 - strlen($level));

        if (!empty($context)) {
            $contextContent = [];
            foreach ($context as $key => $value) {
                if (is_array($value) || is_object($value)) {
                    $value = str_replace("\n", "\n\t\t", print_r($value, true));
                }

                $contextContent[] = sprintf(
                    "\t\t%s%s",
                    self::dark_gray($key, ['bold']),
                    self::dark_gray(': '.$value, null)
                );
            }
            $parsedContext = "\n".implode("\n", $contextContent);
        }

        $formattedMessage = sprintf("[ %s ]%s%s%s\n", $level, $padding, $message, $parsedContext ?? null);

        echo self::$foreground($formattedMessage, $background);
    }
}


